backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,552 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrNoV1             = errors.New("the v1 API is not available for the Kubernetes pipeline store")
+	ErrUnsupportedField = errors.New("the field is unsupported")
+)
+
+const (
+	maxRetries    = 10
+	retryInterval = 10 * time.Second
+)
+
+type PipelineStoreKubernetes struct {
+	client ctrlclient.Client
+}
+
+func NewPipelineStoreKubernetes(k8sClient ctrlclient.Client) *PipelineStoreKubernetes {
+	return &PipelineStoreKubernetes{client: k8sClient}
+}
+
+func (k *PipelineStoreKubernetes) GetPipelineByNameAndNamespaceV1(name string, namespace string) (*model.Pipeline, *model.PipelineVersion, error) {
+	return nil, nil, ErrNoV1
+}
+
+func (k *PipelineStoreKubernetes) GetPipelineByNameAndNamespace(name string, namespace string) (*model.Pipeline, error) {
+	k8sPipeline := v2beta1.Pipeline{}
+
+	err := k.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, &k8sPipeline)
+	if k8serrors.IsNotFound(err) {
+		return nil, util.NewResourceNotFoundError("Namespace/Pipeline", fmt.Sprintf("%v/%v", namespace, name))
+	} else if err != nil {
+		return nil, util.NewInternalServerError(
+			err, "Failed to get a pipeline with name %v and namespace %v", name, namespace,
+		)
+	}
+
+	return k8sPipeline.ToModel(), nil
+}
+
+func (k *PipelineStoreKubernetes) ListPipelinesV1(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, []*model.PipelineVersion, int, string, error) {
+	return nil, nil, 0, "", ErrNoV1
+}
+
+func (k *PipelineStoreKubernetes) ListPipelines(filterContext *model.FilterContext, opts *list.Options) ([]*model.Pipeline, int, string, error) {
+	k8sPipelines := v2beta1.PipelineList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if filterContext.ReferenceKey != nil && filterContext.ReferenceKey.Type == model.NamespaceResourceType {
+		listOptions = append(listOptions, ctrlclient.InNamespace(filterContext.ReferenceKey.ID))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &k8sPipelines, listOptions...)
+	if err != nil {
+		return nil, 0, "", util.NewInternalServerError(
+			err, "Failed to find the pipeline associated with this pipeline version",
+		)
+	}
+
+	pipelines := make([]*model.Pipeline, 0, len(k8sPipelines.Items))
+
+	for _, k8sPipeline := range k8sPipelines.Items {
+		if opts.Filter == nil {
+			pipelines = append(pipelines, k8sPipeline.ToModel())
+			continue
+		} else {
+			found, err1 := opts.Filter.FilterK8sPipelines(k8sPipeline)
+			if err1 != nil {
+				return nil, 0, "", err1
+			}
+			if found {
+				pipelines = append(pipelines, k8sPipeline.ToModel())
+			}
+		}
+	}
+
+	// Because controller-client does not have sorting, use this function to sort by fields.
+	sort.Slice(pipelines, func(i, j int) bool {
```

Comment:
I think we may want to use `sort.SliceStable` here instead (and for other areas of the code).